File: backend/app/schemas/collection_pack.py
"""
采集套餐相关 Schema
定义采集套餐、采集订单的请求/响应数据结构
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ==================== 采集套餐 Schema ====================

class CollectionPackResponse(BaseModel):
    """采集套餐响应"""

    id: UUID
    name: str
    credit_amount: int
    price: float
    description: Optional[str] = None
    is_active: bool

    model_config = {"from_attributes": True}


class CollectionPackCreateRequest(BaseModel):
    """创建采集套餐请求"""

    name: str = Field(..., min_length=1, max_length=100, description="套餐名称")
    credit_amount: int = Field(..., ge=1, description="积分数量（条）")
    price: float = Field(..., ge=0, description="价格（元）")
    description: Optional[str] = Field(None, max_length=255, description="套餐描述")
    is_active: bool = Field(True, description="是否启用")


class CollectionPackUpdateRequest(BaseModel):
    """更新采集套餐请求"""

    name: Optional[str] = Field(None, min_length=1, max_length=100, description="套餐名称")
    credit_amount: Optional[int] = Field(None, ge=1, description="积分数量（条）")
    price: Optional[float] = Field(None, ge=0, description="价格（元）")
    description: Optional[str] = Field(None, max_length=255, description="套餐描述")
    is_active: Optional[bool] = Field(None, description="是否启用")


# ==================== 采集订单 Schema ====================

class CollectionOrderResponse(BaseModel):
    """采集订单响应"""

    id: UUID
    user_id: UUID
    user_name: str
    email: str
    pack_id: UUID
    pack_name: str
    credit_amount: int
    amount: float
    payment_method: str
    status: str
    transaction_id: Optional[str] = None
    paid_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class CreateCollectionOrderRequest(BaseModel):
    """创建采集订单请求"""

    pack_id: UUID = Field(..., description="套餐ID")
    payment_method: str = Field(..., description="支付方式: wechat/alipay")


# ==================== 用户积分余额 Schema ====================

class UserCollectionBalanceResponse(BaseModel):
    """用户积分余额响应"""

    user_id: UUID
    balance: int
    total_purchased: int

    model_config = {"from_attributes": True}

File: src/api/collectionPack.ts
import { api } from '@/lib/api';

// ==================== 类型定义 ====================

export interface CollectionPack {
  id: string;
  name: string;
  credit_amount: number;
  price: number;
  description: string | null;
  is_active: boolean;
}

export interface CollectionOrder {
  id: string;
  user_id: string;
  user_name: string;
  email: string;
  pack_id: string;
  pack_name: string;
  credit_amount: number;
  amount: number;
  payment_method: string;
  status: string;
  transaction_id: string | null;
  paid_at: string | null;
  created_at: string;
}

export interface UserCollectionBalance {
  user_id: string;
  balance: number;
  total_purchased: number;
}

// ==================== 主 API 对象 ====================

export const collectionPackApi = {
  // ==================== 用户端 API ====================

  getPacks: async (): Promise<CollectionPack[]> => {
    const data = await api.get<any>('/v1/collection-packs/plans');
    return data || [];
  },

  purchasePack: async (packId: string, paymentMethod: string = 'wechat'): Promise<CollectionOrder> => {
    const data = await api.post<any, any>('/v1/collection-packs/purchase', {
      pack_id: packId,
      payment_method: paymentMethod,
    });
    return data || {} as CollectionOrder;
  },

  getBalance: async (): Promise<UserCollectionBalance> => {
    const data = await api.get<any>('/v1/collection-packs/balance');
    return data || { user_id: '', balance: 0, total_purchased: 0 };
  },

  getMyOrders: async (page: number = 1, pageSize: number = 20): Promise<CollectionOrder[]> => {
    const data = await api.get<any>(
      `/v1/collection-packs/my-orders?page=${page}&page_size=${pageSize}`
    );
    return data || [];
  },

  // ==================== 后台管理 API ====================

  admin: {
    getPacks: async (): Promise<CollectionPack[]> => {
      const data = await api.get<any>('/v1/admin/collection-packs');
      return data || [];
    },

    createPack: async (data: Partial<CollectionPack>): Promise<CollectionPack> => {
      const result = await api.post<any, any>('/v1/admin/collection-packs', data);
      return result || {} as CollectionPack;
    },

    updatePack: async (id: string, data: Partial<CollectionPack>): Promise<CollectionPack> => {
      const result = await api.put<any, any>(`/v1/admin/collection-packs/${id}`, data);
      return result || {} as CollectionPack;
    },

    deletePack: async (id: string): Promise<void> => {
      await api.delete<any>(`/v1/admin/collection-packs/${id}`);
    },

    getOrders: async (params: {
      user_id?: string;
      status?: string;
      page?: number;
      page_size?: number;
    } = {}): Promise<{ list: CollectionOrder[]; total: number }> => {
      const query = new URLSearchParams();
      if (params.user_id) query.append('user_id', params.user_id);
      if (params.status) query.append('status', params.status);
      if (params.page) query.append('page', String(params.page));
      if (params.page_size) query.append('page_size', String(params.page_size));
      const queryStr = query.toString();
      const url = `/v1/admin/collection-packs/orders${queryStr ? '?' + queryStr : ''}`;
      const data = await api.get<any>(url);
      return data || { list: [], total: 0 };
    },

    updateOrderStatus: async (id: string, newStatus: string): Promise<void> => {
      await api.put<any, any>(
        `/v1/admin/collection-packs/orders/${id}/status`,
        null,
        { params: { new_status: newStatus } }
      );
    },
  },
};
